consensus/propagation/have_wants.go

```diff
@@ -112,11 +113,154 @@ func (blockProp *Reactor) handleHaves(peer p2p.ID, haves *proptypes.HaveParts) {
 		schema.Haves,
 	)
 
+	go blockProp.enqueueRemainingRequests(p, height, round, remaining)
+
 	// keep track of the parts that this node has requested.
 	p.AddRequests(height, round, hc)
+	p.SetRequestCount(currentPeerConcurrentRequests)
 	blockProp.broadcastHaves(haves, peer, int(parts.Total()))
 }
 
+// filterRequests identifies and excludes block parts that have been fully requested or already requested from the peer.
+// returns the parts that couldn't be requested due to the concurrentPerPeerRequestLimit or nil, along with the new concurrent peer requests count.
+// TODO unit test
+// TODO also change this in the catchup!!
+func (blockProp *Reactor) filterRequests(peer p2p.ID, height int64, round int32, hc *bits.BitArray) (*bits.BitArray, int64) {
```

Comment:
we can simplify this significantly by simply passing have parts (or some summary) in a queue to each peer routine, then checking in that routine as recovery parts come in which parts we have and don't have in that queue